### ISSUE TITLE
[PSyIR] C backend and loop indices

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#523]
We need to carefully design the loop indices when outputting PSyIR in C: in typical C the loop should look like:
```
for(i=0; i<n; i+=1)
```
but since the original code is Fortran, loop indices would be expected to be 1 to n (not 0 to n-1). This would matter if e.g. the loop variable is used to compute a value, or in an if statement like:
```
   a(i) = c*i
   if (mod(i,2)==0) ...
```
But if we use a loop like:
```
for(i=1; i<=n; i+=1)
```
then we need to adjust all array accesses. A first loop implementation is done as part of #515, but probably needs to be revisited later.

**Comment by hiker** [comment]
As discussed I've created the branch subtract_one_from_c_arrays_indices with my small change (which fails for gocean1p0_opencl.py).

Just for brainstorming: might it be worth to add a symbolic maths library? We have now potentially index expressions like ``((i-1)+1)*something+...``  (if I understood @sergisiso's suggestion) - and while I would assume a compiler can optimise this, it might be easier to get this sorted out. This library could also be useful for some dependency analysis handling, e.g. if ``a(i+k+1)`` and ``a(i+1+k)`` are accessing the same index (I have added very limited support for this - but it can only handle two elements correctly).

**Comment by rupertford** [comment]
Yes, we discussed this as a possibility just before the meeting but forgot to bring it up. sympy would be an option.

**Comment by hiker** [comment]
That was the name I couldn't remember :) Might be an overkill (and maybe we could make it optional??).
